### DISCUSSION TITLE
ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component. Expected string, but got <spacy.pipeline.sentencizer.Sentencizer object at 0x000001CA1A69DF00> (name: 'None').

### DISCUSSION THREAD

**Discussion by dalalBerrezoug** [explosion/spaCy#7743] [Help: Coding & Implementations] [answered]
i don't understand the error
my code:
def getSentences(text):
    nlp = English()
    nlp.add_pipe(nlp.create_pipe('sentencizer'))

    document = nlp(text)
    return [sent.string.strip() for sent in document.sents]

the error:

   sentences = getSentences(text)
  File "C:\Users\***\PycharmProjects\tutospacy\main.py", line 8, in getSentences
    nlp.add_pipe(nlp.create_pipe('sentencizer'))
  File "C:\Python\Python39\lib\site-packages\spacy\language.py", line 749, in add_pipe
    raise ValueError(err)
ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component. Expected string, but got <spacy.pipeline.sentencizer.Sentencizer object at 0x000001CA1A69DF00> (name: 'None').

- If you created your component with `nlp.create_pipe('name')`: remove nlp.create_pipe and call `nlp.add_pipe('name')` instead.

- If you passed in a component like `TextCategorizer()`: call `nlp.add_pipe` with the string name instead, e.g. `nlp.add_pipe('textcat')`.

- If you're using a custom component: Add the decorator `@Language.component` (for function components) or `@Language.factory` (for class components / factories) to your custom component and assign it a name, e.g. `@Language.component('your_name')`. You can then run `nlp.add_pipe('your_name')` to add it to the pipeline.

my Environment
- **spaCy version:** 3.0.5
- **Platform:** Windows-10-10.0.19041-SP0
- **Python version:** 3.9.4
- **Pipelines:** en_core_web_sm (3.0.0)

**Comment by polm** [comment, marked as answer]
The way `add_pipe` works changed in v3. Your code is using the v2 style, but like the error says you need to change it. Your code should look like this:

    nlp.add_pipe('sentencizer')

**Reply by greybraun** [reply]
Nice to meet you @polm.
Excuse me, can you fix my problem?
I can't compile this code.
                import cov_bsv
                nlp = cov_bsv.load()
got this error:
           ![image](https://user-images.githubusercontent.com/113200672/189400798-89eb383e-8492-4d3c-8c8e-1a9c999511c9.png)

**Reply by polm** [reply]
Please open a new thread, paste errors as text, not images, and explain what `cov_bsv` is.

**Reply by greybraun** [reply]
I wanted to load model on this site.
https://spacy.io/universe/project/cov-bsv

simple code:
          import cov_bsv  # model
          nlp = cov_bsv.load() # load model

error you have seen occoured on second line.

And, thank you for your reply.

**Reply by polm** [reply]
Please [open a new thread](https://github.com/explosion/spaCy/discussions/new).

**Reply by greybraun** [reply]
> Please [open a new thread](https://github.com/explosion/spaCy/discussions/new).
Excuse me, can we discuss on skype? Mine: live:.cid.8eae674c7e1e7d0c

**Comment by Polindem** [comment]


**Reply by polm** [reply]
Uh, not really sure what your error has to do with this one, but if it's a textacy issue maybe you should ask on their repo?

**Comment by Manan-3000** [comment]
Giving me the same kind of error in my code 
def train_model(train_data):
    if 'ner' not in nlp.pipe_names:
        ner = nlp.create_pipe('ner')
        nlp.add_pipe(ner, last = True)
    
    for _, annotation in train_data:
        for ent in annotation['entities']:
            ner.add_label(ent[2])
            
    
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
    with nlp.disable_pipes(*other_pipes):  # only train NER
        optimizer = nlp.begin_training()
        for itn in range(10):
            print("Starting iteration " + str(itn))
            random.shuffle(train_data)
            losses = {}
            index = 0
            for text, annotations in train_data:
                try:
                    nlp.update(
                        [text],  # batch of texts
                        [annotations],  # batch of annotations
                        drop=0.2,  # dropout - make it harder to memorise data
                        sgd=optimizer,  # callable to update weights
                        losses=losses)
                except Exception as e:
                    pass
                
            print(losses)
            
The error:
ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component. 
Expected string, but got <spacy.pipeline.ner.EntityRecognizer object at 0x00000260D46BB040> (name: 'None').

- If you created your component with `nlp.create_pipe('name')`: remove nlp.create_pipe and call `nlp.add_pipe('name')` instead.

- If you passed in a component like `TextCategorizer()`: call `nlp.add_pipe` with the string name instead, e.g. `nlp.add_pipe('textcat')`.

- If you're using a custom component: Add the decorator `@Language.component` (for function components) or `@Language.factory` (for class components / factories) to your custom component and assign it a name, e.g. `@Language.component('your_name')`. You can then run `nlp.add_pipe('your_name')` to add it to the pipeline.

**Reply by svlandeg** [reply]
The error message actually fully explains what you need to do to fix this:

> If you created your component with nlp.create_pipe('name'): remove nlp.create_pipe and call nlp.add_pipe('name') instead.
